theano/tensor/opt.py

```diff
@@ -4129,12 +4143,27 @@ def local_sum_div_dimshuffle(node):
                                          'warn.sum_div_dimshuffle_bug'
                                          ' to False.')
 
-                        new_denom = T.DimShuffle(
-                            thing_dimshuffled.type.broadcastable,
-                            new_new_order)(thing_dimshuffled)
-                    return [T.true_div(node.op(numerator), new_denom)]
-                # else:
-                #    print 'incompatible dims:', axis, new_order
+                    if isinstance(node.op, T.Sum):
+                        op_on_compatible_dims = T.sum(
+                            numerator, axis=compatible_dims)
+                        div_op = T.true_div(
+                            op_on_compatible_dims,
+                            optimized_dimshuffle)
+                        op_on_incompatible_dims = T.sum(
+                            div_op,
+                            axis=reordered_incompatible_dims)
+                    elif isinstance(node.op, T.elemwise.Prod):
+                        op_on_compatible_dims = T.prod(
+                            numerator, axis=compatible_dims)
+                        div_op = T.true_div(
+                            op_on_compatible_dims,
+                            (optimized_dimshuffle **
+                                T.prod([numerator.shape[ax]
+                                        for ax in compatible_dims])))
+                        op_on_incompatible_dims = T.prod(
+                            div_op,
+                            axis=reordered_incompatible_dims)
+                    return [op_on_incompatible_dims]
```

Comment:
@carriepl , Let me explain what I'm doing here by an example:
Let's say we have this:
```python
x = T.dtensor3() 
y = T.dvector()
z = T.prod(x / y.dimshuffle('x', 0, 'x'), axis=(0, 1, 2))
```
There are 3 different things here:
1. ```compatible_dims``` which is ```[0, 2]```
2. ```in_compatible_dims``` which is ```[1]```
3. ```optimized_dimshuffle_order``` which is remaining dimensions in ```dimshuffle``` which has nothing to do with ```prod```. In this case it is ```('x', 0)```
Strategy is to:
1. apply ```prod``` on ```compatible_dims```.
2. divide result by new ```dimshuffle```. (since new order is ```('x', 0)``` it is equal to not applying ```dimshuffle``` so we don't)
3. apply outer ```prod``` on ```in_compatible_dims```. (Note that ```in_compatible_dims``` will change from ```[1]``` to ```[0]```) because of the first ```prod```.

Finally the optimized graph is:
```python
z = T.prod(T.prod(x, axis=(0, 2)) / y ** T.prod(list_of_shapes_of_compatible_dims), axis=0)
```

Here's the ```debugprint``` results before and after optimizations. Can you confirm that?
Before:
```python
Prod{axis=[0, 1, 2], acc_dtype=float64} [@A] ''   2
 |Elemwise{true_div,no_inplace} [@B] ''   1
   |<TensorType(float64, 3D)> [@C]
   |DimShuffle{x,0,x} [@D] ''   0
     |<TensorType(float64, vector)> [@E]
```
After:
```python
Prod{acc_dtype=float64} [@A] ''   7
 |Elemwise{Composite{(i0 / (i1 ** i2))}}[(0, 0)] [@B] ''   6
   |Prod{axis=[0, 2], acc_dtype=float64} [@C] ''   2
   | |<TensorType(float64, 3D)> [@D]
   |<TensorType(float64, vector)> [@E]
   |InplaceDimShuffle{x} [@F] ''   5
     |Prod{acc_dtype=int64} [@G] ''   4
       |MakeVector [@H] ''   3
         |Shape_i{0} [@I] ''   1
         | |<TensorType(float64, 3D)> [@D]
         |Shape_i{2} [@J] ''   0
           |<TensorType(float64, 3D)> [@D]
```